Repo restructure

samaperrin opened the pull request:
# Why have changes been made?

A proper restructure needed to take place before we went any further, as we're about to introduce new models and logic from Phil. 

# What changes have been made?

- Scripts that were previously classed as utils have now been sorted into functions or utils. Those that were useful to others outside of this project and/or used repeatedly throughout the workflow were turned into functions and stored in the functions folder under the subfolder of the pipeline stage (import, integration or modelling) it pertained to. Utils were kept in a utils subfolder under the pipeline directory. 
- README file was updated to reflect the new structure as well. Certain sections of the main scripts (found in the masterScript.R file) were turned into util functions to clean up these scripts.
- A new run was also made with more detailed environmental data to make sure the new folder structure works.

Commit:
Redistribution of util scripts as functions or process-specific util scripts

Commit:
Renaming and redistribution of main source scripts

Commit:
Update of master script to match new folders

Commit:
README in temp folder no longer needed

Commit:
README update with new descriptions of folder structure - still needs work

Commit:
Updates to map data for visualisation

Commit:
Data from new pipeline run

Commit:
Aligning of name change

RRTogunov commented on functions/import/species/defineRegion.R:
```diff
@@ -6,22 +6,7 @@
 
-# Figure out if region and level are supplied. If they are not, check whether a speciesDataList exists in the
-# environment, and use the attributes to automatically find these parameters.
-if (!exists("region") | !exists("level")) {
```
I think we could retain (a version of) this to ensure all variables are properly defined.

RRTogunov commented on functions/import/species/defineRegion.R:
```diff
@@ -6,22 +6,7 @@
-    }
-  }
-}
+defineRegion <- function(level = "county", region = "50", runBuffer = FALSE, extentCoords = NA) {
```
Also, I think this function can be simplified & generalised to 
function(region = "50", level = "country", buffer = 0). where region could either be element within level, or a spatial object (eg sf polygon of interest, sf bbox/terra ext, spatraster, or sf point locations). and Buffer is the distance argument of st_buffer (i.e., default 0 = FALSE as it is now, but otherwise can be any positive value up to practitioner).
I also have code to accept "country" or "continent" for level and obtain polygon for any country. 
I'm happy to implement these changes if they sound good.

RRTogunov commented on pipeline/import/environmentalImport.R:
```diff
@@ -60,9 +66,8 @@ for (parameter in 1:length(selectedParameters)) {
-    # source(paste0("utils/environmentalImport/", focalParameter, ".R"))
     dataSource <- parameters$dataSource[parameters$parameters == focalParameter]
-    source(paste0("utils/environmentalImport/get_", dataSource, ".R"))
+    source(paste0("pipeline/import/utils/defineEnvSource.R"))
```
This could be a candidate to be a general download function in future versions.

RRTogunov commented on functions/import/environment/get_geonorge.R:
```diff
@@ -65,32 +48,26 @@ if (file.exists("data/temp/geonorge/elevationRaster.tiff")) {
-  dirNames <- list.files("data/temp/geonorge/DTM10UTM33", full.names = TRUE)
+  dirNames <- list.files(targetDirExt, full.names = TRUE)
   
   # Aggregate each file to 100m and add to list, then merge
```
to generalise, aggregation could be determined based on target resolution argument (really only needed here to increase computation), otherwise highest resolution should be retained.

RRTogunov commented on functions/import/environment/get_geonorge.R:
```diff
@@ -1,40 +1,26 @@
+# See available data sources: https://nedlasting.geonorge.no/geonorge/Basisdata/
 
-# First check to see if geonorge elevation data has already been imported. If not, import.
+get_geonorge <- function(repo = "Basisdata", dataName = "DTM10UTM33", targetDir) {
```
In future, we should look into generalising this function for other geonorge datasets. In particular, I think converting to terra::spatRaster or sf might need to be different, and may need different code to aggregate/merge tiles/shapes (if they exist).
We could leave function as is and just add warning if they ask for repo or dataName that we have not tested, and we can just add to that list as we test different datasets (same can be done for other downloaders like get_worldclim or get_coppernicus, etc...).

RRTogunov commented on pipeline/import/utils/defineEnvSource.R:
```diff
@@ -0,0 +1,50 @@
+    writeRaster(elevation, "data/temp/geonorge/elevationRaster.tiff", overwrite=TRUE)
+  }
+  
+  # Now get the raster you're actually looking for
```
this could be pulled outside of "geonorge" if a different data source is used for DEM. Could have a separate section for "derived" datasets like slope and aspect that load/download whatever is necessary for specified dataSource.

RRTogunov commented on pipeline/integration/utils/speciesRichnessConversion.R:
```diff
@@ -3,34 +3,13 @@
-regionGeometry <- readRDS(paste0("data/run_", dateAccessed, "/regionGeometry.RDS"))
-
 
 regionGeometry_buffer <- vect(st_buffer(regionGeometry, 1))
```
st_buffer looks to be called a few times (eg. environmentalImport script adds 20km buffer), should we specify buffer once at the start or is there merit to using different buffers for different steps of pipeline? I'm not sure if it is necessary here.

RRTogunov commented on functions/models/modelPreparation.R:
```diff
@@ -0,0 +1,68 @@
+    
+    # Initialise workflow, creating folder for model result storage
+    workflow <- startWorkflow(
+      Projection = '+proj=utm +zone=32 +ellps=WGS84 +datum=WGS84 +units=m +no_defs',
```
We should probably define working projection/resolution somewhere in master script and use that throughout. I receive errors/warnings with certain things being in wrong projection or no covariates in mesh, which I think is tied to multiple crs definitions and transformations in several parts of the workflow.

RRTogunov commented on functions/models/modelPreparation.R:
```diff
@@ -0,0 +1,68 @@
+  workflowList <- list()
+  
+  # Begin running different species groups
+  for (i in 1:length(focalTaxa)) {
```
This can probably be popped out as a separate function; called once if one focal taxa, or using lapply if multiple.

RRTogunov commented on functions/models/modelPreparation.R:
```diff
@@ -0,0 +1,68 @@
+    focalGroup <- focalTaxa[i]
+    focalGroupSpecies <- focalSpecies$species[focalSpecies$taxonomicGroup %in% focalGroup]
+    
+    focalSpeciesData <- speciesData
```
It looks like speciesData is only defined outside of function 'modelPreparation'? if so, it should be added as argument in function definition.

RRTogunov committed:
Update masterScript.R

correct path for species data processing script

RRTogunov committed:
update geonorge download

make target directory for geonorge if it is missing and use tempFolderName for as temporary target directory.

RRTogunov committed:
Update speciesRichnessConversion.R

convert script to use terra package instead of raster.

RRTogunov committed:
Update meshTest.R

define target CRS and use throughout function.

RRTogunov committed:
Update modelPreparation.R

remove unused resolution argument from area definition.

RRTogunov committed:
Update speciesModelRuns.R

update folder name of environmental covariates

Commit:
Added functions imports
